source/endpoints_ruleset.c

```diff
@@ -0,0 +1,939 @@
+            if (s_parse_endpoints_rule_data_error(wrapper->allocator, error_node, &rule.rule_data.error)) {
+                AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Failed to extract error rule data.");
+                goto error_clean_up;
+            }
+            break;
+        }
+        case AWS_ENDPOINTS_RULE_TREE: {
+            if (s_parse_endpoints_rule_data_tree(wrapper->allocator, value, &rule.rule_data.tree)) {
+                AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Failed to extract tree rule data.");
+                goto error_clean_up;
+            }
+            break;
+        }
+        default:
+            AWS_FATAL_ASSERT(false);
+    }
+
+    /* Optional fields */
+    struct aws_json_value *documentation_node =
+        aws_json_value_get_from_object(value, aws_byte_cursor_from_c_str("documentation"));
+    if (documentation_node != NULL) {
+        struct aws_byte_cursor documentation_cur;
+        if (aws_json_value_get_string(documentation_node, &documentation_cur)) {
+            AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Failed to extract parameter documentation.");
+            goto error_clean_up;
+        }
+        rule.documentation = aws_string_new_from_cursor(wrapper->allocator, &documentation_cur);
+    }
+
+    aws_array_list_push_back(wrapper->array, &rule);
+
+    return AWS_OP_SUCCESS;
+
+error_clean_up:
+    aws_endpoints_rule_clean_up(&rule);
+    return aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_PARSE_FAILED);
+}
+
+static int s_init_ruleset_from_json(
+    struct aws_allocator *allocator,
+    struct aws_endpoints_ruleset *ruleset,
+    struct aws_byte_cursor json) {
+    AWS_PRECONDITION(allocator);
+    AWS_PRECONDITION(ruleset);
+    AWS_PRECONDITION(aws_byte_cursor_is_valid(&json));
+
+    /*
+     * TODO: this could be more efficient. currently user allocates mem for json
+     * string, we create a copy in json parser and then string fields create
+     * another copy.
+     */
+
+    struct aws_json_value *root = aws_json_value_new_from_string(allocator, json);
+
+    if (root == NULL) {
+        AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Failed to parse provided string as json.");
+        return aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_PARSE_FAILED);
+    }
+
+    struct aws_byte_cursor version_cur;
+    struct aws_json_value *version_node = aws_json_value_get_from_object(root, aws_byte_cursor_from_c_str("version"));
+    if (version_node == NULL || aws_json_value_get_string(version_node, &version_cur)) {
+        AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Failed to extract version.");
+        aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_UNSUPPORTED_RULESET);
+        goto error_clean_up;
+    }
+
+#ifdef VERSION_CHECK /* TODO: samples are currently inconsistent with versions. skip check for now */
+    if (aws_byte_cursor_eq_c_str(&version_cur, &s_supported_version)) {
```

Comment:
should this be case-sensitive? or not?

Reply:
case sensitive is fine for now. version strings will be 2 or 3 numbers separated by dots. so caseness does not matter